### PR TITLE
Feature: Wave Config for Targets

### PR DESCRIPTION
**Pull request by StewartW**
*Issue #, if available:* #290

*Description of changes:*
Introduction of an optional wave configuration in the target schema. 

This contains a configurable int that defines the wave size. This wave size is then used to create batches inside the target structure class. 

These batches are then passed into the cdk stacks and accessed in a similar fashion to how targets were already accessed. 
This PR doesn't have tests yet, as I noticed that there's no test for pipeline generation beyond Source / Build. 

The default batch size is set to 50 (Max actions per stage) and in the default adf_pipeline_type, it will create one stage per wave, allowing for OUs with more than 50 accounts to be seperated over multiple stages. 

Also added in an exclude property to the target for removing accounts by Id.  (issue: #145 )
Also fixed an issue where if an account shared a tag with an OU and that tag was used as a Target it would break.  (Issue: #296 )

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_default_pipeline_type.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+
+# pylint: skip-file
+
+import pytest
+from pprint import pprint
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_default_pipeline_type.py:7]: This line is not used I think.

```suggestion
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_default_pipeline_type.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+    build_stage = code_pipeline["Properties"]["Stages"][1]
+    build_stage_action = build_stage["Actions"][0]
+    assert build_stage_action["ActionTypeId"]["Category"] == "Build"
+    assert build_stage_action["ActionTypeId"]["Owner"] == "AWS"
+    assert build_stage_action["ActionTypeId"]["Provider"] == "CodeBuild"
+
+    assert len(build_stage["Actions"]) == 1
+
+    region_name = "eu-central-1"
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_default_pipeline_type.py:64]: It looks like another test kicks-off from this line.
Could you add its function/test definition?

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_default_pipeline_type.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_default_pipeline_type.py:179]: Same here, looks like the function def is missing. As another test setup kicks-off from here.

**Review by sbkok — CHANGES_REQUESTED** [review]
Thank you for adding this!
Only a few minor comments to fix. I'm looking forward to merge this one.

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_default_pipeline_type.py (hunk 3 above)
Comment (StewartW) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_default_pipeline_type.py:76]: Think this was just a bad copy paste job, it's the same test from the pipeline_creation test

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/schema_validation.py [hunk 4]
```diff
@@ -302,7 +302,7 @@
     int
 )]
 
-TARGET_WAVE_CONFIG_SCHEME = {
+TARGET_wave_SCHEME = {
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/schema_validation.py:305]:
```suggestion
TARGET_WAVE_SCHEME = {
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/schema_validation.py [hunk 5]
```diff
@@ -317,7 +317,7 @@
     Optional("properties"): Or(CODEBUILD_PROPS, JENKINS_PROPS, CLOUDFORMATION_PROPS, CODEDEPLOY_PROPS, S3_DEPLOY_PROPS, SERVICECATALOG_PROPS, LAMBDA_PROPS, APPROVAL_PROPS),
     Optional("regions"): REGION_SCHEMA,
     Optional("exclude", default=[]): [str],
-    Optional("wave_config", default={"size": 50}): TARGET_WAVE_CONFIG_SCHEME
+    Optional("wave", default={"size": 50}): TARGET_wave_SCHEME
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/schema_validation.py:320]:
```suggestion
    Optional("wave", default={"size": 50}): TARGET_WAVE_SCHEME
```

**Review by sbkok — DISMISSED** [review]
Thank you! This looks great.

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_default_pipeline_type.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+# pylint: skip-file
+
+from aws_cdk import core
+from cdk_stacks.main import PipelineStack
+
+
+def test_pipeline_creation_outputs_as_expected_when_input_has_1_target_with_2_waves():
+    region_name = "eu-central-1"
+    acount_id = "123456789012"
```
Comment (deltagarrett) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_default_pipeline_type.py:12]: Can we fix this typo?

Comment (deltagarrett) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_default_pipeline_type.py:12]: Others are where this is used.

**Review by deltagarrett — CHANGES_REQUESTED** [review]
Very minor fix to address.

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_default_pipeline_type.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+        "properties": {"account_id": "123456789012"},
+    }
+    stack_input["input"]["default_providers"]["build"] = {
+        "provider": "codebuild",
+        "properties": {"account_id": "123456789012"},
+    }
+
+    stack_input["ssm_params"][region_name] = {
+        "modules": "fake-bucket-name",
+        "kms": f"arn:aws:kms:{region_name}:{acount_id}:key/my-unique-kms-key-id",
```
Comment (deltagarrett) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_default_pipeline_type.py]: Typo

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_default_pipeline_type.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    assert len(target_1_wave_1["Actions"]) == 3
+
+    target_1_wave_2 = code_pipeline["Properties"]["Stages"][3]
+    assert target_1_wave_2["Name"] == "deployment-stage-1-wave-1"
+    assert len(target_1_wave_2["Actions"]) == 3
+
+
+def test_pipeline_creation_outputs_as_expected_when_input_has_2_targets_with_2_waves_and_1_wave():
+    region_name = "eu-central-1"
+    acount_id = "123456789012"
```
Comment (deltagarrett) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_default_pipeline_type.py]: Typo

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_default_pipeline_type.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+        "properties": {"account_id": "123456789012"},
+    }
+    stack_input["input"]["default_providers"]["build"] = {
+        "provider": "codebuild",
+        "properties": {"account_id": "123456789012"},
+    }
+
+    stack_input["ssm_params"][region_name] = {
+        "modules": "fake-bucket-name",
+        "kms": f"arn:aws:kms:{region_name}:{acount_id}:key/my-unique-kms-key-id",
```
Comment (deltagarrett) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_stacks/tests/test_default_pipeline_type.py]: Typo

**Review by deltagarrett — CHANGES_REQUESTED** [review]
Minor typos to fix.

**Review by deltagarrett — APPROVED** [review]
LGTM

**Review by sbkok — APPROVED** [review]
LGTM!

**Comment by AngeloGelmini** [comment]
When I launch the aws-deployment-framework-pipelines I get this error:
`
Traceback (most recent call last):
  File "/codebuild/output/src433857646/src/adf-build/cdk/generate_pipeline_inputs.py", line 194, in <module>
    main()
  File "/codebuild/output/src433857646/src/adf-build/cdk/generate_pipeline_inputs.py", line 190, in main
    thread.join()
  File "/codebuild/output/src433857646/src/adf-build/python/thread.py", line 30, in join
    raise self.exc
  File "/codebuild/output/src433857646/src/adf-build/python/thread.py", line 20, in run
    self.ret = self._target(
  File "/codebuild/output/src433857646/src/adf-build/cdk/generate_pipeline_inputs.py", line 117, in worker_thread
    pipeline.template_dictionary["targets"].append(target.target_structure.generate_waves())
AttributeError: 'dict' object has no attribute 'target_structure'
`

My deployment map is:
`
pipelines:
  - name: parameters
    default_providers:
      source:
        provider: codecommit
        properties:
          account_id: 111111111111111 # id source account
      build:
        provider: codebuild
      deploy:
        provider: cloudformation
        properties:
          action: replace_on_failure
    params:
      notification_endpoint: ...
    targets: # accounts where to deploy resources
      - target: 99999999999 # deployment account
        name: name-
        regions: eu-central-1
`

**Comment by sbkok** [comment]
Hi @AngeloGelmini, thanks for reporting this.
I have fixed this issue together with Stewart, I will open the PR for that fix shortly.

I'm sure you are aware that Wave Config is not released as part of ADF yet. Therefore we don't recommend relying on that yet, as it might be (or actually is as you noticed) unstable.

We are actively debugging the code to make sure that the different features and changes introduced for the next release work together before we release ADF v3.2.0.

**Pull request merged**